### PR TITLE
Add shared preferences functionality

### PR DESCRIPTION
**Pull request by SofiaZarzabal**
## Description
Add shared preference option with load/save data functionality.

## Evidence
Screenshare - 2022-12-26 10_17_06 PM

### CODE REVIEW

File: app/src/main/res/layout/activity_info_manager.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:id="@+id/til_name"
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_info_manager.xml]: change this id by name_container, please

File: app/src/main/res/layout/activity_info_manager.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginHorizontal="@dimen/size_info_activity_manager_margin_horizontal"
+        android:layout_marginTop="@dimen/size_info_activity_manager_margin_top"
+        android:hint="@string/til_name_text_hint"
+        app:layout_constraintTop_toTopOf="parent">
+
+        <com.google.android.material.textfield.TextInputEditText
+            android:id="@+id/tiet_name"
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_info_manager.xml]: change this ID by name_edit_text, please

File: app/src/main/java/com/example/datastoreapp/storetype/SPStore.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import android.content.SharedPreferences
+
+class SPStore(private val context: Context) {
+    private val storage: SharedPreferences = context.getSharedPreferences(SP_NAME, 0)
+
+    fun saveName(name: String) {
+        storage.edit().putString(SP_USERNAME, name).apply()
+    }
+
+    fun getName(): String = storage.getString(SP_USERNAME, "username") ?: ""
```
Comment (SartiMau) [inline app/src/main/java/com/example/datastoreapp/storetype/SPStore.kt]: do not hardcode, please put the string "username" and "" into a constant.

File: app/src/main/java/com/example/datastoreapp/activity/MainActivity.kt [hunk 4]
```diff
@@ -16,17 +17,16 @@ class MainActivity : AppCompatActivity() {
...
         setListeners()
     }
 
     private fun updateUI(data: MainData) {
-        when (data.state) {
-            GO_TO_SHARED_PREFERENCES_VIEW -> {} //start Activity() to shared preferences
-            GO_TO_PROTO_DATASTORE_VIEW -> TODO() //start Activity() to proto datastore
-            GO_TO_PREFERENCES_DATASTORE_VIEW -> TODO() //start Activity() to preferences datastore
-        }
+        this.startActivity(Intent(this, InfoManagerActivity::class.java).apply {
```
Comment (SartiMau) [inline app/src/main/java/com/example/datastoreapp/activity/MainActivity.kt]: From the viewModel, you still posting the states GO_TO_SHARED_PREFERENCES_VIEW, GO_TO_PROTO_DATASTORE_VIEW, and GO_TO_PREFERENCES_DATASTORE_VIEW.
I recommend keeping the states here too

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/datastoreapp/activity/MainActivity.kt (hunk 4 above)
Comment (SartiMau) [reply app/src/main/java/com/example/datastoreapp/activity/MainActivity.kt]: Also, create a getIntent method into the companion of the InfoManagerActivity that response with the correspondent intent

File: app/src/main/java/com/example/datastoreapp/fragment/SharedPreferencesFragment.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        super.onCreateView(inflater, container, savedInstanceState)
+        binding = FragmentSharedPreferencesBinding.inflate(layoutInflater)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        viewModel = SharedPreferencesViewModel(SharedPreferencesStore(requireContext()))
```
Comment (SartiMau) [inline app/src/main/java/com/example/datastoreapp/fragment/SharedPreferencesFragment.kt]: be careful with the `requireContext()` as if it comes null you will get a crash, it is better if you ask for the context directly

File: app/src/main/java/com/example/datastoreapp/storetype/SharedPreferencesStore.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        storage.edit().putString(SP_USERNAME, name).apply()
+    }
+
+    fun getName(): String = storage.getString(SP_USERNAME, USERNAME) ?: EMPTY
+
+    companion object {
+        const val SP_NAME = "SPStore"
+        const val SP_USERNAME = "SP_USERNAME"
+        const val USERNAME = "USERNAME"
+        const val EMPTY = ""
```
Comment (SartiMau) [inline app/src/main/java/com/example/datastoreapp/storetype/SharedPreferencesStore.kt]: change this to EMPTY_STRING to be more representative

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -7,11 +7,21 @@
...
+        android:id="@+id/btn_go_to_info_manager"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/size_main_activity_button"
+        android:text="Go to info manager"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <!--<Button
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_main.xml]: Why do we have all these commented code?

Comment (SartiMau) [reply app/src/main/res/layout/activity_main.xml]: a puppy died when you made this :P

File: app/src/main/res/layout/fragment_preferences_data_store.xml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="Preferences data store"
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_preferences_data_store.xml]: don't hardcode, use string res

File: app/src/main/res/layout/fragment_proto_data_store.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="Proto data store"
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_proto_data_store.xml]: don't hardcode, use string res

File: app/src/main/res/layout/fragment_shared_preferences.xml [hunk 10]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="Shared preferences"
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_shared_preferences.xml]: don't hardcode, use string res

**Review by SartiMau — CHANGES_REQUESTED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
LG

**Pull request merged**